### PR TITLE
ParallelDescriptor::ReduceReal wrappers

### PR DESCRIPTION
**Pull request by WeiqunZhang**
## Summary

Make ParallelDescriptor::ReduceReal wrappers template on the floating point
data type.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

File: Src/Base/AMReX_ParallelDescriptor.H [hunk 1]
```diff
@@ -1120,6 +1189,333 @@ Scatter(T* /*t*/, size_t /*n*/, const T1* /*t1*/, size_t /*n1*/, int /*root*/)
...
+					  MyTeam().get_lead_comm()) );
+	}
+	BL_MPI_REQUIRE( MPI_Bcast(recv.dataPtr(), cnt, Mpi_typemap<T>::type(),
+				  0, MyTeam().get_team_comm()) );
+    }
+    else
+#endif
+    {
+	BL_MPI_REQUIRE( MPI_Allreduce(r,
+				      recv.dataPtr(),
```
Comment (sayerhs) [inline Src/Base/AMReX_ParallelDescriptor.H]: @WeiqunZhang I am wondering if there is a reason why AMReX doesn't use `MPI_IN_PLACE` here?

Comment (WeiqunZhang) [reply Src/Base/AMReX_ParallelDescriptor.H]: Thanks for pointing this out.  We should `MPI_IN_PLACE` here.  Will update it.

**Review by atmyers — APPROVED** [review]

**Comment by atmyers** [comment]
Do think the test failure is related to `MPI_IN_PLACE`?

**Comment by WeiqunZhang** [comment]
Yes, I think it's fixed now.

**Comment by WeiqunZhang** [comment]
For `MPI_Reduce`, `MPI_IN_PLACE` can only be used on the root process.

**Pull request merged**